### ISSUE TITLE
[Performance] [Audit] `ThumbnailImage` triggers layout shifts, impeding FlatList performance

### ISSUE THREAD

**Issue by jsamr** [Expensify/App#4115]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

This report is part of #3957, [scenario "Rendering Individual chat messages"](https://github.com/Expensify/Expensify.cash/issues/3957#issuecomment-881045715).

### Flamegraph

*The full commit log can be inspected with Flipper or React Devtools, see https://github.com/Expensify/Expensify.cash/issues/3957#issuecomment-881045715 for instructions.*

You can witness the layout shift behavior in the below Framegraph (c244, select `VirtualizedList` component)
![image](https://user-images.githubusercontent.com/3646758/125970578-849fbe5d-1981-40b1-866e-1752eed1f7aa.png)

The four commits (c222, c244, c266 and c280) are caused by `last` state of `VirtualizedList` component being changed, this is the signature of layout shifts. There are exactly 3 images in the report. So we have 1 initial commit + 3 layout shifts. **Those 3 layout shifts span over 1.3s, this is considerable.**

To get more into the details of `VirtualizedList` internals, a layout shift causes [`_onCellLayout`](https://github.com/facebook/react-native/blob/ff9c8cdac9b1734ad5f193116880b7b91e08bc6e/Libraries/Lists/VirtualizedList.js#L1272) callback to be invoked,
which calls [`_scheduleCellsToRenderUpdate`](https://github.com/facebook/react-native/blob/ff9c8cdac9b1734ad5f193116880b7b91e08bc6e/Libraries/Lists/VirtualizedList.js#L1642) which calls [` _updateCellsToRender`](https://github.com/facebook/react-native/blob/ff9c8cdac9b1734ad5f193116880b7b91e08bc6e/Libraries/Lists/VirtualizedList.js#L1730), which in turn causes `last` state to change.

### Cause

https://github.com/Expensify/Expensify.cash/blob/76c3157a41d86beb89c8cf107a68573280c06ce7/src/components/ThumbnailImage.js#L59-L63

The `ThumbnailImage` component first draws a 100x100 box to display the image, then waits for physical dimensions to be available and redraws the image with the newfound dimensions. This causes the `FlatList` cell layout to shift, and requires `FlatList` to recompute the position of each cell. This can have a sensible impact on performance if a discussion has many image attachments, since every shift will cause the `FlatList` to recompute the position of each cell, potentially re-rendering some cells.

### Proposal: store images physical dimensions in markup

Store the physical dimensions of the image as attributes (`width`, `height`) of the `<img>` element. Thus, we can render the image one and only once and avoid those costly shifts.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @stephanieelliott (`AutoAssignerTriage`), see https://stackoverflow.com/c/expensify/questions/4749 for more details.

**Comment by marcaaron** [comment]
>This can have a sensible impact on performance if a discussion has many image attachments, since every shift will cause the FlatList to recompute the position of each cell, potentially re-rendering some cells.

Oh wow, yeah this sounds like it could be pretty bad depending on how costly it is to recompute the position. I feel like we should be able to test the impacts with a very simple test case where 2 chats are switched back and forth and timed. e.g. one will have 500 regular comments the other has 499 comments and 1 attachments (or more) then we compare the difference.

>Store the physical dimensions of the image as attributes (width, height) of the <img> element. Thus, we can render the image one and only once and avoid those costly shifts.

Open to hearing thoughts on how we can store this information. It would be nice if there is some way to [do this at the server level](https://github.com/Expensify/Web-Expensify/blob/64f8a88897fe89ace3180758cd6923c86c416469/lib/ReportUtils.php#L294-L302) when generating the HTML for the image. Otherwise I think we will need to cache the size or delay rendering the `FlatList` until we have the sizes available. cc @thienlnam

**Comment by thienlnam** [comment]
Very interesting findings!

> I feel like we should be able to test the impacts with a very simple test case where 2 chats are switched back and forth and timed. e.g. one will have 500 regular comments the other has 499 comments and 1 attachments (or more) then we compare the difference.

To add to this I'm pretty sure we have some kind of internal caching going on so this is probably more apparent the first time the chat loads as consecutive loads of the chat with images will be much speedier.  Additionally, I am curious is this shifts all cells, or just visible cells around a resized image

> It would be nice if there is some way to do this at the server level when generating the HTML for the image. 

So we can store the image sizes when we generate the HTML, but before we render the image we still need to scale the image so that it is relative to the screen size.  Additionally, when a user resizes the window we'd want to account for that and show the image resizing as well instead of showing a white screen until the resize is done.  Something I'm thinking of that we could do is in conjunction to storing the full image size in the HTML, we have some ONYX.KEY store the device dimensions and then when rendering the image we do some scaling of the size inside the <img> directly and so it only renders a single time with that specific size

**Comment by kidroca** [comment]
> Additionally, I am curious is this shifts all cells, or just visible cells around a resized image
I'm pretty sure this would affect the position of every cell in the virtualized portion that is loaded, but I'm not certain it will recompute/re-render cells besides the ones containing images that change

The reason is we don't implement `getItemLayout` which means we leave the FlatList to compute how much height a cell takes and then calculate it's absolute position (offset) in the list. To do that the list has to render the cells.
After the cells are rendered if some change (the image containing ones) the list doesn't have to re-render the rest, but only change their offset position which should be quick, but maybe not quick enough

**Comment by jsamr** [comment]
@thienlnam 
> before we render the image we still need to scale the image so that it is relative to the screen size

Just want to point out that this scaling behavior is already implemented in react-native-render-html internal Image component. You can reuse the internal component in the custom renderer [such as described here](https://meliorence.github.io/react-native-render-html/docs/content/images#by-reusing-internal-building-blocks); just replace `useIMGElementState` with `useIMGElementStateWithCache` since you would have dimensions beforehand (and thus no `"loading"` state). Final note: the max width of the scaled image will be derived from [`computeEmbeddedMaxWidth`](https://meliorence.github.io/react-native-render-html/api/renderhtmlconfig#computeembeddedmaxwidth) prop.

**Comment by thienlnam** [comment]
Oh nice! In that case we should just be able to store the height/width with the image when we generate it and remove a lot of the `imageWithSizeCalculation` code

**Comment by MelvinBot** [comment]
Triggered auto assignment to @robertjchen (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @NicMendonca (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by NicMendonca** [comment]
Posted to Upwork: https://www.upwork.com/jobs/~018fc7af43553297f3

**Comment by MelvinBot** [comment]
Triggered auto assignment to @deetergp (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by mananjadhav** [comment]
One of the solutions could be replacing `Image` with `react-native-fast-image`. 

One shortcoming with FastImage is that it doesn't give an API to fetch the image dimensions. I've used it for large lists with thumbnails and it works very well with improved performance with caching.

**Comment by thienlnam** [comment]
@mananjadhav If it isn't able to fetch the image dimensions, it would run into the same issue with layout shifts after the image size is calculated right? Also, Image seems to have some internal caching, what performance benefits would we get with `react-native-fast-image`?

Side note: I think this needs to be an internal issue (at least until we start storing image dimensions in Auth)

**Comment by mananjadhav** [comment]
@thienlnam Yeah. I was reading through the chats that we'll start storing the dimensions. That + Cached Image will give a huge performance boost. It isn't ideal to fetch `Image.getSize` for all the images. 

I've pointed another issue where Image flicker is caused because we don't have cached images.

**Comment by marcaaron** [comment]
Just want to state that I have seen no evidence that `react-native-fast-image` will provide any huge performance boosts for this app. I've set it up a few times and there was not really any noticeable changes. But it takes 5 minutes to set up if someone can create a benchmark to prove the value.

**Comment by haditariq** [comment]
My solution would be to compress images on the backend and have multiple variants of each image in the database, for example having 3 sizes of different sizes would make loading easy and quick with this we can cache the chat in the app. At first, it will take some time to load and from next it would be really quick.

**Comment by quinthar** [comment]
I completely agree that we should pregenerate thumbnails made to the exact
sizes we are going to show.

I think one way to do this is to generate them on the fly if they don't
already exist. So when you request the image from our cdn, which is routed
through our servers for authentication purposes, you should request what
thumbnail size you want. The first time we get the request we won't have it
so we'll generate it and then save it for the future. And then the second
time it will already be there.

This allows a different platforms to request thumbnails of different sizes,
but the second person on the same platform will get an optimized experience.

**Comment by RushirajJhala** [comment]
Agree, its good option to generate thumbnail image in backend whenever any image is uploaded , keep 2 copies , thumbnail and original and use  one as per need.

**Comment by MelvinBot** [comment]
@deetergp, @NicMendonca it looks like no one is assigned to work on this job.
Please double the price or add a comment stating why the job isn't being doubled.

**Comment by NicMendonca** [comment]
Doubled price: https://www.upwork.com/jobs/~01416052ece8f86c7c

**Comment by thienlnam** [comment]
I think we have a clear idea to move forward on this project:

Internal: (From @quinthar's comment here https://github.com/Expensify/App/issues/4115#issuecomment-889228242)
Generate automatic thumbnail/image sizes on request, and cache/save those so that they are ready when requested another time.  

External: 
Update the image component to request a specific sized image based on the app dimensions and use that size initially so that the image isn't resized after loading

**Comment by thienlnam** [comment]
Since the external portion is pending on the internal, I'm going to remove the exported tag from this and we can re-add it once we have the internal part completed

**Comment by haseebb** [comment]
Hi

I will use  react-native-fast-image.

I've used it for large lists with thumbnails and it works very well with improved performance with caching.

**Comment by thienlnam** [comment]
Hello @haseebb,

There's been a bit of discussion about `react-native-fast-image` above but especially [here](https://github.com/Expensify/App/issues/4115#issuecomment-887838308).  It would be great to see if there was any performance difference in the first place.

**Comment by haseebb** [comment]
Hi @thienlnam 

Can you plz send a screen recording of the app while the app is loading the images.
I want to see the loading delay.

**Comment by thienlnam** [comment]
@haseebb Feel free to follow these steps https://github.com/Expensify/App#local-development to get set up with a development environment.  You should be able to play around with images there

**Comment by deetergp** [comment]
I've created [this GH](https://github.com/Expensify/Expensify/issues/173677) for our engineers to handle the back-end portion of this effort.

**Comment by NicMendonca** [comment]
@thienlnam add me back if we need this added to Upwork 😃

**Comment by deetergp** [comment]
There has been some [discussion](https://github.com/Expensify/Expensify/issues/173677#issuecomment-900185404) about the backend for this earlier in the week. I need to spend some time digging it more, but have been a bit distracted this week. Hoping to have some time to deal with it early next.

**Comment by deetergp** [comment]
This is continuing to hold on an internal engineer picking up the API side of this issue.

**Comment by deetergp** [comment]
Still no internal takers on https://github.com/Expensify/Expensify/issues/173677. I may pick it up before week's end if nobody beats me to it.

**Comment by deetergp** [comment]
Holding for N6

**Comment by deetergp** [comment]
Still no takers on the internal GH.

**Comment by deetergp** [comment]
No progress to speak of.

**Comment by deetergp** [comment]
Still no internal takers on https://github.com/Expensify/Expensify/issues/173677

**Comment by deetergp** [comment]
That internal GH has has been closed by Melvin. There have been some other discussions in the contributor sphere around ways to speed up image performance that may address the findings from this audit, so maybe by the next time it gets marked overdue we can just close it.

**Comment by kidroca** [comment]
I see we now have `react-native-fast-image` installed which might help with caching. We're also planning on caching images in Onyx. Caching might solve the problem, but only for consecutive loads 

When you switch to different device you'll probably experience the problem again, or when you're cache expires

Adding the width and height to the `img` tag (as suggested) should address the layout shift problem

**Comment by deetergp** [comment]
Agreed, @kidroca. Going to close this since we've taken differing approaches to speeding up image handling.
